sdk/identity/azure-identity/inc/azure/identity/managed_identity_credential.hpp

```diff
@@ -42,6 +42,91 @@ namespace Azure { namespace Identity {
     std::string ToString() const { return m_resourceId; }
   };
 
+  /**
+   * @brief The type of managed identity identifier depending on how the managed identity is
+   * configured.
+   *
+   * @remark This can either be system assigned, or user assigned which corresponds to an identifier
+   * that represents either client ID, resource ID, or object ID, depending on how the managed
+   * identity is configuerd.
+   */
+  enum class ManagedIdentityIdType
+  {
+    SystemAssigned,
+    ClientId,
+    ObjectId,
+    ResourceId,
+  };
+
+  /**
+   * @brief The type of managed identity and its corresponding identifier.
+   *
+   * @remark This property holds the type and unique identifier of either a system or user-assigned
+   * managed identity.
+   */
+  class ManagedIdentityType final {
+  private:
+    ManagedIdentityIdType m_idType;
+    std::string m_id;
+
+  public:
+    /**
+     * @brief Constructs the type of managed identity.
+     *
+     * @remark This defaults to ManagedIdentityIdType::SystemAssigned.
+     */
+    explicit ManagedIdentityType() : m_idType(ManagedIdentityIdType::SystemAssigned) {}
+
+    /**
+     * @brief Constructs the type of managed identity.
+     *
+     * @param idType The type of the managed identity identifier.
+     * @param id The value of the managed identity identifier. This can be either a client ID,
+     * resource ID, or object ID.
+     *
+     * @remark For ManagedIdentityIdType::SystemAssigned, the id must be an empty string.
+     *
+     * @remark Make sure the type of ID matches the value of the ID. For example, the client
+     * ID and object ID are NOT interchangeable, even though they are both Uuid values.
+     */
+    explicit ManagedIdentityType(ManagedIdentityIdType idType, std::string id = std::string())
```

Comment:
Why default `id` to an empty string? Is this to enable the case where you explicitly specify managed identity type?

I'd rather there be only one way of specifying the system assigned managed identity - right now, you can do it two different ways which seems weird.

Reply:
Gotcha. Makes sense.
Will throw on `ManagedIdentityIdType::SystemAssigned`, then, for this case.

Reply:
Actually, even if I make std::string id required, the user can still pass in an empty string.

So, should we continue to allow `SystemAssigned` be passed in? I don't know if it is worth blocking.

Otherwise, I can throw in all cases, if an empty string is provided.

Reply:
I left the validation as is. If the user passes in a valid `ManagedIdentityIdType` value of `SystemAssigned`, with an empty id value string, that's fine.